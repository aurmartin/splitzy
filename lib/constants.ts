export const primaryHue = 200;

export const Colors = {
  primaryDark: `hsl(${primaryHue} 80% 40%)`,
  primary: `hsl(${primaryHue} 40% 50%)`,
  primaryLight: `hsl(${primaryHue} 40% 80%)`,

  secondary: `hsl(0 0% 70%)`,
  secondaryLight: `hsl(0 0% 80%)`,

  red: `hsl(0 50% 40%)`,

  green: `hsl(120 50% 40%)`,

  background: `hsl(0 0% 94%)`,
};
