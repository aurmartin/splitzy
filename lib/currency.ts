import { type Dinero, type Currency } from "dinero.js";

const MINOR_UNIT_FACTOR: Partial<Record<Currency, number>> = {
  EUR: 100,
  USD: 100,
};

const getAmountColor = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "hsl(120 70% 30%)";
  if (amount.getAmount() < 0) return "hsl(0 70% 30%)";
  return "black";
};

const getAbsoluteAmount = (amount: Dinero): Dinero => {
  return amount.getAmount() < 0 ? amount.multiply(-1) : amount;
};

const getBalanceMessage = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "On vous doit";
  if (amount.getAmount() < 0) return "Vous devez";
  return "Vous êtes à jour";
};

const getMinorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount * factor;
};

const getMajorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount / factor;
};

export {
  getAmountColor,
  getAbsoluteAmount,
  getBalanceMessage,
  getMinorUnitAmount,
  getMajorUnitAmount,
};
