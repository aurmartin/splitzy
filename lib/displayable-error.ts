export class DisplayableError extends Error {
  title: string;
  message: string;

  constructor(title: string, message: string) {
    super(`${title}: ${message}`);
    this.title = title;
    this.message = message;
  }
}
