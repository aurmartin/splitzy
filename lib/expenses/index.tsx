export * from "./types";
export * from "./hooks";

export { ExpensesRepository } from "./repository";
export { ExpenseService } from "./service";
