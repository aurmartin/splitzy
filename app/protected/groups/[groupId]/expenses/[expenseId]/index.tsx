import Label from "@/components/label";
import { ListGroup } from "@/components/list-group";
import { Screen } from "@/components/screen";
import { useSnackBar } from "@/components/snack-bar";
import { useSystem } from "@/components/system-provider";
import { Text } from "@/components/text";
import { TopBar } from "@/components/top-bar";
import { TopBarAction, TopBarDeleteAction } from "@/components/top-bar-action";
import {
  ExpenseService,
  useAmounts,
  useExpense,
  type Expense,
} from "@/lib/expenses";
import { getLocale } from "@/lib/locale";
import { router, useLocalSearchParams } from "expo-router";
import { ScrollView, View } from "react-native";

interface ShowExpenseScreenProps {
  groupId: string;
  expense: Expense;
}

function ShowExpenseScreen({ groupId, expense }: ShowExpenseScreenProps) {
  const snackBar = useSnackBar();
  const system = useSystem();
  const amounts = useAmounts(expense);

  const handleDelete = async () => {
    try {
      await ExpenseService.deleteExpense(system, expense.id);

      router.back();

      snackBar.show("Dépense supprimée", "success");
    } catch (error) {
      console.error("Delete expense error", error);

      snackBar.show("Une erreur est survenue", "error");
    }
  };

  const handleEdit = () => {
    router.replace(`/protected/groups/${groupId}/expenses/${expense.id}/edit`);
  };

  return (
    <Screen>
      <TopBar
        rightActions={
          <>
            <TopBarDeleteAction onPress={handleDelete} />
            <TopBarAction iconName="pencil" onPress={handleEdit} />
          </>
        }
      />

      <Text type="headlineMedium" style={{ marginBottom: 16 }}>
        {expense.title}{" "}
        {expense.splitExpense.total.setLocale(getLocale()).toFormat()}
      </Text>

      <Text style={{ marginBottom: 16 }}>
        Payé par {expense.payerName} le{" "}
        {new Date(expense.createdAt).toLocaleDateString()}
      </Text>

      <ScrollView
        showsVerticalScrollIndicator={false}
        alwaysBounceVertical={false}
      >
        <Label>Répartition</Label>
        <ListGroup style={{ marginBottom: 16 }}>
          {Object.entries(amounts).map(([name, amount]) => (
            <View key={name} style={{ justifyContent: "space-between" }}>
              <Text key={name}>{name}</Text>
              <Text>{amount.setLocale(getLocale()).toFormat()}</Text>
            </View>
          ))}
        </ListGroup>

        {expense.receipt && (
          <>
            <Label>Ticket de caisse</Label>
            <ListGroup itemHeight={70}>
              {expense.receipt.items.map((item) => (
                <View key={item.description}>
                  <View style={{ gap: 4, flex: 1, marginRight: 16 }}>
                    <Text
                      style={{ flexShrink: 1 }}
                      numberOfLines={1}
                      ellipsizeMode="middle"
                    >
                      {item.humanReadableDescription}
                    </Text>
                    <Text
                      type="bodyMedium"
                      style={{ color: "hsl(0, 0%, 40%)" }}
                    >
                      {item.description}
                    </Text>
                  </View>

                  <Text>{item.price.setLocale(getLocale()).toFormat()}</Text>
                </View>
              ))}
            </ListGroup>
          </>
        )}
      </ScrollView>
    </Screen>
  );
}

export default function ShowExpenseRoute() {
  const { expenseId, groupId }: { expenseId: string; groupId: string } =
    useLocalSearchParams();

  const expense = useExpense(expenseId);

  if (!expense) {
    return null;
  }

  return <ShowExpenseScreen groupId={groupId} expense={expense} />;
}
