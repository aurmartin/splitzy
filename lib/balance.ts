import dinero, { type Dinero } from "dinero.js";
import { type Expense, ExpenseService, useExpenses } from "@/lib/expenses";
import { type Group } from "@/lib/groups";

export type Balance = Record<string, Dinero>;

export type Reimbursement = {
  from: string;
  to: string;
  amount: Dinero;
};

export function computeBalance(group: Group, expenses: Expense[]): Balance {
  const balance = group.members.reduce((acc, member) => {
    acc[member] = dinero({ amount: 0, currency: group.currency });
    return acc;
  }, {} as Balance);

  expenses.forEach((expense) => {
    const payer = expense.payerName;
    const total = expense.splitExpense.total;

    balance[payer] = balance[payer].add(total);

    const amounts = ExpenseService.getAmounts(expense);

    Object.entries(amounts).forEach(([name, amount]) => {
      balance[name] = balance[name].subtract(amount);
    });
  });

  return balance;
}

export const useBalance = (group: Group): Balance => {
  const expenses = useExpenses(group.id);
  return computeBalance(group, expenses);
};

export function computeOptimalReimbursements(
  balance: Balance,
): Reimbursement[] {
  const reimbursements: Reimbursement[] = [];
  const members = Object.keys(balance);

  // Create a copy of the balance object
  const workingBalance = members.reduce((acc, member) => {
    acc[member] = balance[member];
    return acc;
  }, {} as Balance);

  // Continue while there are non-zero balances
  while (members.some((member) => !workingBalance[member].isZero())) {
    // Find the person who owes the most (most negative balance)
    const maxDebtor = members.reduce((a, b) =>
      workingBalance[a].lessThan(workingBalance[b]) ? a : b,
    );

    // Find the person who is owed the most (most positive balance)
    const maxCreditor = members.reduce((a, b) =>
      workingBalance[a].greaterThan(workingBalance[b]) ? a : b,
    );

    // If both balances are effectively zero (less than 1 cent), break
    if (
      workingBalance[maxDebtor].isZero() &&
      workingBalance[maxCreditor].isZero()
    ) {
      break;
    }

    // Calculate the reimbursement amount (minimum of absolute values)
    const debtorAmount = workingBalance[maxDebtor].multiply(-1); // Make positive
    const creditorAmount = workingBalance[maxCreditor];
    const amount = debtorAmount.lessThan(creditorAmount)
      ? debtorAmount
      : creditorAmount;

    // Update the balances
    workingBalance[maxDebtor] = workingBalance[maxDebtor].add(amount);
    workingBalance[maxCreditor] = workingBalance[maxCreditor].subtract(amount);

    // Record the reimbursement
    reimbursements.push({
      from: maxDebtor,
      to: maxCreditor,
      amount,
    });
  }

  return reimbursements;
}
