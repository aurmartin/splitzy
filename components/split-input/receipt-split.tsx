import { Text } from "@/components/text";
import { primaryHue } from "@/lib/constants";
import { type Receipt, type ReceiptSplit } from "@/lib/expenses";
import { convertSplit } from "@/components/split-input/convert";
import { getLocale } from "@/lib/locale";
import React from "react";
import { TouchableOpacity, View } from "react-native";

const ReceiptSplitInput = (props: {
  value: ReceiptSplit;
  onChange: (value: ReceiptSplit) => void;
}) => {
  const { value, onChange } = props;

  const isSelected = (item: Receipt["items"][number], member: string) => {
    return item.paid_for.includes(member);
  };

  const toggleSelect = (itemIndex: number, member: string) => {
    onChange({
      ...value,
      receipt: {
        ...value.receipt,
        items: value.receipt.items.map((item, index) => {
          if (index === itemIndex) {
            if (item.paid_for.includes(member)) {
              return {
                ...item,
                paid_for: item.paid_for.filter((m) => m !== member),
              };
            }

            return {
              ...item,
              paid_for: [...item.paid_for, member],
            };
          }
          return item;
        }),
      },
    });
  };

  const backgroundColor = (item: Receipt["items"][number], member: string) => {
    return isSelected(item, member)
      ? `hsl(${primaryHue}, 100%, 90%)`
      : `hsl(0, 0%, 90%)`;
  };

  const amountSplit = convertSplit(value);

  return (
    <View
      style={{
        flex: 1,
        backgroundColor: "white",
        borderRadius: 4,
      }}
    >
      <View
        style={{
          gap: 8,
          padding: 12,
          borderBottomWidth: 1,
          borderColor: "hsl(0, 0%, 80%)",
        }}
      >
        {Object.entries(amountSplit.amounts).map(([member, amount]) => (
          <View
            key={member}
            style={{
              display: "flex",
              flexDirection: "row",
              justifyContent: "space-between",
            }}
          >
            <Text numberOfLines={1}>{member}</Text>
            <Text>{amount.setLocale(getLocale()).toFormat()}</Text>
          </View>
        ))}
      </View>

      <View style={{ padding: 12, gap: 8 }}>
        {value.receipt.items.map((item, index) => (
          <View key={index} style={{ gap: 8 }}>
            <View
              style={{
                flex: 1,
                flexDirection: "row",
                justifyContent: "space-between",
                alignItems: "center",
              }}
            >
              <View>
                <Text>{item.humanReadableDescription}</Text>
                <Text type="bodyMedium" style={{ color: "hsl(0, 0%, 40%)" }}>
                  {item.description}
                </Text>
              </View>
              <Text>{item.price.setLocale(getLocale()).toFormat()}</Text>
            </View>

            <View style={{ flex: 1, flexDirection: "row", gap: 8 }}>
              {value.members.map((member) => (
                <TouchableOpacity
                  key={member}
                  style={{
                    backgroundColor: backgroundColor(item, member),
                    padding: 8,
                    borderRadius: 8,
                  }}
                  onPress={() => toggleSelect(index, member)}
                >
                  <Text>{member}</Text>
                </TouchableOpacity>
              ))}
            </View>
          </View>
        ))}
      </View>
    </View>
  );
};

function createReceiptSplit(receipt: Receipt, members: string[]): ReceiptSplit {
  return {
    type: "receipt",
    total: receipt.total,
    members,
    receipt,
  };
}

export { ReceiptSplitInput, createReceiptSplit };
