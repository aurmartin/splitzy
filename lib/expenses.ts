import { convertSplit } from "@/components/split-input";
import { useSystem } from "@/components/system-provider";
import { ValidationError } from "@/lib/validation-error";
import { ExpensesRepository } from "@/lib/expenses-repository";
import { System } from "@/lib/system";
import { generateId } from "@/lib/utils";
import { Currency, Dinero } from "dinero.js";
import { useCallback } from "react";

interface BaseSplit {
  type: "percentage" | "amount" | "equal" | "receipt";
  total: Dinero;
  members: string[];
}

interface PercentageSplit extends BaseSplit {
  type: "percentage";
  ratios: { [member: string]: number };
}

interface EqualSplit extends BaseSplit {
  type: "equal";
}

interface AmountSplit extends BaseSplit {
  type: "amount";
  amounts: { [member: string]: Dinero };
}

interface ReceiptSplit extends BaseSplit {
  type: "receipt";
  receipt: Receipt;
}

type Split = PercentageSplit | EqualSplit | AmountSplit | ReceiptSplit;

interface Receipt {
  date: Date;
  total: Dinero;
  title: string;
  currency: Currency;

  items: {
    description: string;
    humanReadableDescription: string;
    price: Dinero;
    paid_for: string[];
  }[];
}

interface Expense {
  id: string;
  groupId: string;
  title: string;
  payerName: string;
  splitExpense: Split;

  createdAt: Date;
  updatedAt: Date;
  deletedAt: Date | undefined;

  receipt: Receipt | null;
}

interface ExpenseCreateParams {
  groupId: Expense["groupId"];
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
}

interface ExpenseUpdateParams {
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
}

const useAddExpense = () => {
  const system = useSystem();

  return useCallback(
    (params: ExpenseCreateParams) => addExpense(system, params),
    [system],
  );
};

const useEditExpense = () => {
  const system = useSystem();
  return useCallback(
    (id: string, params: ExpenseUpdateParams) =>
      editExpense(system, id, params),
    [system],
  );
};

const useDelExpense = () => {
  const system = useSystem();
  return useCallback(
    (id: string) => ExpensesRepository.deleteExpense(system, id),
    [system],
  );
};

const addExpense = async (system: System, params: ExpenseCreateParams) => {
  const validParams = validateCreateExpenseParams(params);
  const expense = createExpense(validParams);
  await ExpensesRepository.insertExpense(system, expense);
};

const editExpense = async (
  system: System,
  id: string,
  params: ExpenseUpdateParams,
) => {
  const expense = ExpensesRepository.getExpense(system, id);
  if (expense === null) {
    throw new Error("Expense not found");
  }

  const validParams = validateUpdateExpenseParams(params);
  const updatedExpense = modifyExpense(expense, validParams);
  await ExpensesRepository.updateExpense(system, updatedExpense);
};

const createExpense = (validParams: ExpenseCreateParams): Expense => {
  return {
    ...validParams,
    id: generateId(),
    createdAt: new Date(),
    updatedAt: new Date(),
    deletedAt: undefined,
  };
};

const modifyExpense = (
  expense: Expense,
  params: ExpenseUpdateParams,
): Expense => {
  return {
    ...expense,
    ...params,
    updatedAt: new Date(),
  };
};

const validateCreateExpenseParams = (params: ExpenseCreateParams) => {
  const errors: Record<string, string> = {};

  if (params.title.length === 0) {
    errors.title = "Veuillez entrer un titre";
  }

  if (params.payerName.length === 0) {
    errors.payerName = "Veuillez sélectionner un membre";
  }

  if (Object.keys(errors).length > 0) {
    throw new ValidationError(errors);
  }

  return params;
};

const validateUpdateExpenseParams = (params: ExpenseUpdateParams) => {
  const errors: Record<string, string> = {};

  if (params.title.length === 0) {
    errors.title = "Veuillez entrer un titre";
  }

  if (params.payerName.length === 0) {
    errors.payerName = "Veuillez sélectionner un membre";
  }

  return params;
};

const getAmountSplit = (expense: Expense): AmountSplit => {
  return convertSplit(expense.splitExpense);
};

const useExpense = ExpensesRepository.useExpense;
const useExpenses = ExpensesRepository.useExpenses;

export type {
  AmountSplit,
  EqualSplit,
  Expense,
  PercentageSplit,
  Receipt,
  ReceiptSplit,
  Split,
};

export {
  addExpense,
  getAmountSplit,
  useAddExpense,
  useDelExpense,
  useEditExpense,
  useExpense,
  useExpenses,
};
