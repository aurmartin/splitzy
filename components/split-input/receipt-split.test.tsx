import * as React from "react";
import { render, screen, userEvent } from "@testing-library/react-native";
import { ReceiptSplitInput, createReceiptSplit } from "./receipt-split";
import dinero from "dinero.js";
import { type Receipt } from "@/lib/expenses";

describe("ReceiptSplit", () => {
  const mockReceipt: Receipt = {
    items: [
      {
        description: "Item_1",
        humanReadableDescription: "Item 1",
        price: dinero({ amount: 1000, currency: "USD" }),
        paid_for: ["user1"],
      },
      {
        description: "Item_2",
        humanReadableDescription: "Item 2",
        price: dinero({ amount: 2000, currency: "USD" }),
        paid_for: [],
      },
    ],
    total: dinero({ amount: 3000, currency: "USD" }),
    date: new Date(),
    title: "Receipt 1",
    currency: "USD",
  };

  describe("ReceiptSplitInput", () => {
    it("renders correctly", () => {
      const mockSplit = {
        type: "receipt" as const,
        total: mockReceipt.total,
        members: ["user1", "user2"],
        receipt: mockReceipt,
      };

      render(<ReceiptSplitInput value={mockSplit} onChange={() => {}} />);
      expect(screen.toJSON()).toMatchSnapshot();
    });

    it("displays all receipt items", () => {
      const mockSplit = {
        type: "receipt" as const,
        total: mockReceipt.total,
        members: ["user1", "user2"],
        receipt: mockReceipt,
      };

      render(<ReceiptSplitInput value={mockSplit} onChange={() => {}} />);

      screen.getByText("Item 1");
      screen.getByText("Item 2");
    });

    it("handles item selection toggle", async () => {
      const user = userEvent.setup();

      const mockSplit = {
        type: "receipt" as const,
        total: mockReceipt.total,
        members: ["user1", "user2"],
        receipt: mockReceipt,
      };

      const mockOnChange = jest.fn();
      render(<ReceiptSplitInput value={mockSplit} onChange={mockOnChange} />);

      // Find and click user2's button for the first item
      const user2Buttons = screen.getAllByText("user2");
      await user.press(user2Buttons[1]);

      expect(mockOnChange).toHaveBeenCalledWith(
        expect.objectContaining({
          receipt: expect.objectContaining({
            items: mockReceipt.items.map((item, index) => ({
              ...item,
              paid_for: index === 0 ? ["user1", "user2"] : [],
            })),
          }),
        }),
      );
    });
  });

  describe("createReceiptSplit", () => {
    it("creates receipt split correctly", () => {
      const members = ["user1", "user2"];

      const result = createReceiptSplit(mockReceipt, members);

      expect(result.type).toBe("receipt");
      expect(result.total).toBe(mockReceipt.total);
      expect(result.members).toEqual(members);
      expect(result.receipt).toBe(mockReceipt);
    });

    it("preserves existing paid_for assignments", () => {
      const members = ["user1", "user2"];

      const result = createReceiptSplit(mockReceipt, members);

      expect(result.receipt.items[0].paid_for).toEqual(["user1"]);
      expect(result.receipt.items[1].paid_for).toEqual([]);
    });
  });
});
