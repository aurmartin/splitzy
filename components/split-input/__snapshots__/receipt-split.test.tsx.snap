// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiptSplit ReceiptSplitInput renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 4,
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "borderBottomWidth": 1,
        "borderColor": "hsl(0, 0%, 80%)",
        "gap": 8,
        "padding": 12,
      }
    }
  >
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
          ]
        }
      >
        user1
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
          ]
        }
      >
        $10.00
      </Text>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
          ]
        }
      >
        user2
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
          ]
        }
      >
        $0.00
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "gap": 8,
        "padding": 12,
      }
    }
  >
    <View
      style={
        {
          "gap": 8,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
              ]
            }
          >
            Item 1
          </Text>
          <Text
            style={
              [
                {
                  "fontSize": 14,
                  "fontWeight": "400",
                  "lineHeight": 20,
                },
                {
                  "color": "hsl(0, 0%, 40%)",
                },
              ]
            }
          >
            Item_1
          </Text>
        </View>
        <Text
          style={
            [
              {
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
            ]
          }
        >
          $10.00
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "hsl(200, 100%, 90%)",
              "borderRadius": 8,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
              ]
            }
          >
            user1
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "hsl(0, 0%, 90%)",
              "borderRadius": 8,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
              ]
            }
          >
            user2
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "gap": 8,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
              ]
            }
          >
            Item 2
          </Text>
          <Text
            style={
              [
                {
                  "fontSize": 14,
                  "fontWeight": "400",
                  "lineHeight": 20,
                },
                {
                  "color": "hsl(0, 0%, 40%)",
                },
              ]
            }
          >
            Item_2
          </Text>
        </View>
        <Text
          style={
            [
              {
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
            ]
          }
        >
          $20.00
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "hsl(0, 0%, 90%)",
              "borderRadius": 8,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
              ]
            }
          >
            user1
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "hsl(0, 0%, 90%)",
              "borderRadius": 8,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
              ]
            }
          >
            user2
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
